[Snackbar] Added new component

HofmannZ opened the pull request:
Hey just created the Snackbar component, it works really good with the bottom navigation component. and follows the [guidelines](https://material.io/guidelines/components/snackbars-toasts.html), still want the implement the multiline version but we can roll out the one line version first.

Please feel free to come up with improvements, I'm checking for PRs at least twice a day.

PS @xotahal could you explain how I get the button styles for this component from getThemes function?

xotahal commented on src/Snackbar/Snackbar.react.js:
```diff
@@ -0,0 +1,207 @@
+};
+
+function getStyles(props, context) {
+    const { snackbar } = context.uiTheme;
```
`const { snackBar, button, buttonFlat, buttonDisabled, buttonRaised } = context.uiTheme;`

Is that what you want?

HofmannZ replied:
Ah, we could do that, but this button should never be raised nor disabled.

xotahal replied:
It's true probably, but I think it doesn't matter ... we'll use flat design ... we don't have to use another styles of button ... And I think it's always better to reuse one code then make duplicity ..

-----
And I think there will be someone in future who wants raised button on snackbar :D

xotahal commented on src/Snackbar/Snackbar.react.js:
```diff
@@ -0,0 +1,207 @@
+        container: View.propTypes.style,
+        message: View.propTypes.style,
+        actionContainer: View.propTypes.style,
+        actionText: View.propTypes.style,
```
We should prefer to reuse the button component here. So, there could be something like:

`button: Button.propTypes.style` and then you can pass it through props to button ..

xotahal commented on src/Snackbar/Snackbar.react.js:
```diff
@@ -0,0 +1,207 @@
+        if (actionText && (typeof actionHandler === 'function')) {
+            return (
+                <Button
+                    style={{ container: styles.actionContainer, text: styles.actionText }}
```
`style={styles.button}` ... unfortunately, there is no way how to change the button text I think .. :/ But it's an issue in button component, not here ..

xotahal commented on src/Snackbar/Snackbar.react.js:
```diff
@@ -0,0 +1,207 @@
+class Snackbar extends PureComponent {
+    constructor(props, context) {
+        super(props, context);
+        const styles = getStyles(props, context);
```
I like it man :)

I've been thinking about this for couple of weeks :) And I want to make a revision for all components .. Unfortunately, you will have a problem with this, since you don't care about changes of style prop in `componentWillReceiveProps` .. 

I was thinking about some method like `shouldStyleUpdate` .. returns true if any property that can affect style was chaged .. (style, size in IconToggle, etc.)

HofmannZ replied:
Yeah, that would be cool for now, I will use componentWillReceiveProps. Let me know when you're going to refactor, I can help you with it.

xotahal commented on src/styles/getTheme.js:
```diff
@@ -397,6 +398,31 @@ export default function getTheme(theme, ...more) {
+        snackbar: StyleSheet.create(merge({
+            container: {
+                flexDirection: 'row',
+                height: 48,
```
It should be constant in `spacing.js`, I think .. snackbarHeight?

xotahal commented on src/styles/getTheme.js:
```diff
@@ -397,6 +398,31 @@ export default function getTheme(theme, ...more) {
+                flexDirection: 'row',
+                height: 48,
+                alignItems: 'center',
+                backgroundColor: '#323232',
```
Take it from color.js

xotahal commented on src/styles/getTheme.js:
```diff
@@ -397,6 +398,31 @@ export default function getTheme(theme, ...more) {
+            },
+            message: {
+                flex: 1,
+                fontSize: 14,
```
Take it from `typography.js`

xotahal commented on src/styles/getTheme.js:
```diff
@@ -397,6 +398,31 @@ export default function getTheme(theme, ...more) {
+                flex: 1,
+                fontSize: 14,
+                marginVertical: 14,
+                color: '#ffffff',
```
color.js

xotahal commented on src/styles/getTheme.js:
```diff
@@ -397,6 +398,31 @@ export default function getTheme(theme, ...more) {
+            },
+            actionText: {
+
+            },
```
These two should be merged - button: { }

HofmannZ commented on src/Snackbar/Snackbar.react.js:
```diff
@@ -68,16 +67,18 @@ function getStyles(props, context) {
+                local.button.text,
+                props.style.button.text,
+            ],
+        },
```
Does not work because `container` is not a valid style property.

xotahal replied:
But it's not because of this, it's because you create button style in `getTheme.js`, isn't it? In this case, we should take default setting directly from `context.uiTheme.button` and merge it with, `snackBar.propTypes.button.style`. Should work, right?

HofmannZ replied:
Yeah, I was thinking about that, but we also need to remove the horizontal padding and set the height to 14dp on the action. Because as described in the guidelines, there is only a 24dp margin on the left and right side of the action text.

We can't add those styles locally because we want to give other developers the option change it globally in their theme. We would eighter need a to add styles for the action and merge it like so:
```js
button: [
  context.uiTheme.button,
  { container: context.uiTheme.snackbar.action },
  local.button,
  props.button.style,
]
```
Or we need to find and nicer way to merge those action button styles with the default button styles.

HofmannZ commented on src/Snackbar/Snackbar.react.js:
```diff
@@ -171,7 +178,7 @@ class Snackbar extends PureComponent {
             return (
                 <Button
-                    style={{ container: styles.actionContainer, text: styles.actionText }}
+                    style={styles.button}
```
As you said it does not work. Changing it back.

HofmannZ commented:
Could you explain me how to get those button stylings in correctly, and how do you want to pass the props for the button just next to the props from the Snackbar or inside an object?

xotahal commented:
Probably, we should pass just single `button` prop and there is also style prop for button, right?

```
Snackbar.propTypes = {
   ...
   button: Button.propTypes,
   ...
}
```
 And then you can do just `<Button {...this.props.button} />`. But also, we should provide some default behaviour. We could add some "sugar" prop for button text because user always wants to change the text of action. So there could be your `actionText` prop for button's text and then please rename `actionHandler` to `onActionPress` - because I use this `on...Press` in every component ... Everything else will be passed via `props.button`.

```<Button {...this.props.button} text={this.props.actionText} onPress={this.props.onActionPress} />```

About style:
There should be some style for button in `getTheme.js` - under `snackBar` as well. Then you can take it from context and combine it with `this.props.button.style`. Is it clear?

What do you think?

HofmannZ commented:
If we add some style for the button under `snackbar` like [this](https://github.com/xotahal/react-native-material-ui/pull/81/commits/1000d32213c7249d8fddfe1b070cda008da5275c), we will get the following error:
```sh
Invariant Violation: "container" is not a valid style property.
StyleSheet button: {
    "container": {
        "height": 14,
        "paddingHorizontal": 0
    }
}
```
So can't put it under there, that's why I think we could do `snackbarAction` inside `getTheme.js`.

xotahal commented:
I've just left a comment on the [old pull request](https://github.com/xotahal/react-native-material-ui/pull/81/commits/1000d32213c7249d8fddfe1b070cda008da5275c):

> But it's not because of this, it's because you create button style in getTheme.js, isn't it? In this case, we should take default setting directly from context.uiTheme.button and merge it with, snackBar.propTypes.button.style. Should work, right?

HofmannZ commented:
@xotahal just read trough it and left a reply.

HofmannZ commented:
Apparently, you can't use the backticks in a zsh terminal. I'm sorry for the last commit message it should say the following:
Added an extra if statement to avoid a possible bug where `container` of `text` inside `button.style` does not exist. Fixed a warning where the required button prop `text` was undefined. Fixed a bug where the button styles were not properly overwritten.

HofmannZ commented:
I think it's all ready now

xotahal commented:
I've just lost it here :) There are 42 commits ... but I still see kind of problems there .. I'll try to fix it ..

HofmannZ commented:
I'm sorry, I use your demo app with my own repo as a submodule, that's why the're so much commits. About the problems, yeah I just know there are some issues, but I had to ship an App with the snack at component. I will continue to add some features to snack bar, such as: message queue (allowing to send multiple messages in a sequence), show the message until the snack bar has completely moved out of the screen (most likely fixed by the event queue)
